File: solutions/Container-With-Most-Water/Container-With-Most-Water.py


class Solution:
    # @return an integer
    def maxArea(self, height):
        head=0
        tail=len(height)-1
        ans=0
        while head<tail:
            tempans=min(height[head],height[tail])*(tail-head)
            ans=max(tempans,ans)
            if height[head]<height[tail]:
                head+=1
            else:
                tail-=1
        return ans
 

File: solutions/Container-With-Most-Water/Container-With-Most-Water.cpp

// class Solution {
// public:
//     int maxArea(vector<int> &height) {
//         int ans=0;
//         int n=height.size();
//         int l=0;
//         int r=n-1;
//         while(l<r)
//         {
//             ans=max(ans,(r-l)*min(height[l],height[r]));
//             if(height[l]<height[r])
//                 l++;
//             else
//                 r--;
//         }
//         return ans;
//     }
// };


 // 20190714
 // O(n)
class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0;
        int r = height.size() - 1;
        int ans = 0;
        while (l < r)
        {
            ans = max(ans, (r - l) * min(height[l], height[r]));
            if (height[l] < height[r])
            {
                l++;
            }
            else
            {
                r--;
            }
        }
        return ans;
    }
};